### PR TITLE
ID generation to better ensure concurrent requests generate unique sequences

### PR DESCRIPTION
**Pull request by boydc7**
Existing ID generation is pretty much guaranteed to result in large numbers of duplicate/repeating ID values (see the [unit test included here](https://github.com/DataDog/dd-trace-dotnet/compare/chad/randomIds#diff-8d036c391380b976d1ea9eb94dc3974c) as proof).  Abstracting the ID generation out to allow easy follow-on implementation if we change it/come up with a better way.

### CODE REVIEW

File: src/Datadog.Trace/ExtensionMethods/AsyncExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+using System;
+using System.Collections.Generic;
+
+namespace Datadog.Trace.ExtensionMethods
+{
+    internal static class AsyncExtensions
+    {
+        internal static T GetOrSet<T>(this AsyncLocalCompat<T> source, Func<T> setter, T unsetValue = default(T))
```
Comment (lucaspimentel) [inline src/Datadog.Trace/ExtensionMethods/AsyncExtensions.cs]: Any reason why this is an extension method and not just an instance method on `AsyncLocalCompat()`?

File: src/Datadog.Trace/Services/RandomIdProvider.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+            _randomProvider = randomProvider ?? throw new ArgumentNullException(nameof(randomProvider));
+        }
+
+        public static RandomIdProvider Instance { get; } = new RandomIdProvider(SimpleDependencyFactory.RandomProvider());
+
+        public ulong GetUInt63Id()
+        {
+            // From https://stackoverflow.com/a/677390
+            var buffer = new byte[sizeof(ulong)];
+
```
Comment (lucaspimentel) [inline src/Datadog.Trace/Services/RandomIdProvider.cs]: Just a formatting nit. I tend to use more empty lines than most C# developers (at least more than @bmermet and @dd-caleb), but even this seems a bit excessive to me.

File: src/Datadog.Trace/Services/ThreadLocalNewRandomProvider.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+using System;
+using System.Threading;
+using Datadog.Trace.Interfaces;
+
+namespace Datadog.Trace.Services
+{
+    internal class ThreadLocalNewRandomProvider : IRandomProvider
```
Comment (lucaspimentel) [inline src/Datadog.Trace/Services/ThreadLocalNewRandomProvider.cs]: Are we using this anywhere? Seems like even a static `Random` (shared process-wide across threads) would be preferable over this implementation. If it's not used, feel free to remove it.

**Comment by boydc7** [comment]
@lucaspimentel well, last night after I PR'd this I kept thinking...and there's another problem.  Basically by using new Random() objects on each thread/async context, no matter how well we seed them, we're still just kicking the can a little further down the road - since Random is seeded with a 32-bit int, we're still guaranteed to wind up with duplicates after a while (and not all that long really in terms of what we use it for I imagine).  The implementations I PR'd are marginally better probably than the original, but not much on that front.

So, I added a singleton-random provider that is not thread-safe but is interlocked on access during ID generation - we wind up with the pros of the new approaches above (i.e. better randomness, no time-depenedent IDs, etc.), plus we have a single Random for the lifetime of the process that will most likely be distinct in terms of random sequence generation from other machines.  The downside is we have to lock the GET off the next value from the random object.

In theory this should result in poorer performance, but on my tests locally it's actually a bit faster - I don't expect that to hold true in a bigger environment where there's more contention, but it's encouraging that it isn't all that bad under normal use.

So, I'd say at this point we just have to pick our approach - once we decide, I can remove the other non-used implementations (they're left right now for purposes of simply being able to run tests locally with different providers if you like).  In the end we might keep a couple (the AsyncLocal implementation and that chosen one maybe or something), but others can just go bye-bye

File: src/Datadog.Trace/ExtensionMethods/AsyncExtensions.cs (hunk 1 above)
Comment (boydc7) [reply src/Datadog.Trace/ExtensionMethods/AsyncExtensions.cs]: Generally you'd ideally keep implementations/interfaces as clean/minimal (i.e. only require implementation of what is needed) and extend from there (which typically allows you to add new/different concrete implementations more easily).  In this case it's not an interface but an actual concrete object, so it'd only really matter if we ever derived it for example.

File: src/Datadog.Trace/Services/RandomIdProvider.cs (hunk 2 above)
Comment (boydc7) [reply src/Datadog.Trace/Services/RandomIdProvider.cs]: I agree, but our linter thing keeps yelling at me...probably only on some of these and not all, but once it started yelling I just started hitting enter.... I'll go see if maybe I can figure out the specifics, but I'm with you here if we can get that thing to stop erroring...

File: src/Datadog.Trace/Services/ThreadLocalNewRandomProvider.cs (hunk 3 above)
Comment (boydc7) [reply src/Datadog.Trace/Services/ThreadLocalNewRandomProvider.cs]: Yea, see below - I only left the extra impelmentations for now so you could run/test/view locally differences if you want, once we decide on the implementation we want to use, I'll get rid of the others

**Comment by bmermet** [comment]
A bit late to the party, but I don't understand why we need so much complexity for random id generation, why don't we go the way Zipkin4Net is doing: https://github.com/openzipkin/zipkin4net/blob/master/Src/zipkin4net/Src/Utils/RandomUtils.cs. The referenced SO post seems to provide a reasonable solution: [link](https://stackoverflow.com/questions/19270507/correct-way-to-use-random-in-multithread-application) with a threadlocal + unique seed for each thread. Also there really is only one place where we need uniqueness on random generation for SpanId and TraceId, in `AgentWriterBuffer.cs` any kind of randomness should be good enough, which makes me a bit worried of introducing so many abstractions `IRandomProvider`, `IIdProvider`, `SimpleDependencyFactory` for this.

**Pull request closed without merging**